### ISSUE TITLE
Access to private Apple APIs prevents acceptance to App Store

### ISSUE THREAD

**Issue by jbelkins** [awslabs/aws-crt-swift#206]
### Describe the bug

The following ticket was filed by a customer on AWS SDK for Swift:  https://github.com/awslabs/aws-sdk-swift/issues/1204

Since the source of this issue is likely cryptography components in `aws-crt-swift`, I am filing this ticket with the original report from the customer.

---

### Describe the bug

We are using the AWS SDK to access an S3 bucket on our iOS app. Since the latest Xcode update we are getting a "Non-public API usage" error when we try to upload our iOS app to Test Flight: 

```
ITMS-90338: Non-public API usage - The app references non-public symbols in {PROJECT_NAME}: _CCCryptorGCMAddAAD, _CCCryptorGCMAddIV, _CCCryptorGCMFinal, _CCCryptorGCMFinalize, _CCCryptorGCMSetIV. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/
```

We have identified that after removing the AWS SDK from our app, we stopped getting this warning. 

### Expected Behavior

The AWS SDK shouldn't use any private Apple APIs to apps can be submitted to Test Flight and the App Store. 

### Current Behavior

We are get the following error when we submit our app to Test Flight: 

```
ITMS-90338: Non-public API usage - The app references non-public symbols in {PROJECT_NAME}: _CCCryptorGCMAddAAD, _CCCryptorGCMAddIV, _CCCryptorGCMFinal, _CCCryptorGCMFinalize, _CCCryptorGCMSetIV. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/
```

Here is a screenshot of the Xcode search for one of the restricted APIs: 
![Screenshot 2023-10-30 at 17 01 56](https://github.com/awslabs/aws-sdk-swift/assets/47433973/332ccb60-f067-4941-a38d-9728ab2a97b3)

### Reproduction Steps

Integrate the SDK into any iOS app and try to access an object from a private S3 bucket (using IAM credentials). 

### AWS SWIFT SDK version used

0.28.0

### Compiler and Version used

Xcode 15.0, swift-driver version: 1.87.1 Apple Swift version 5.9 

### Operating System and version

iOS, any version

### Expected Behavior

App including `aws-sdk-swift` and its dependencies should be accepted for upload by App Store Connect without warning messages or errors.

### Current Behavior

App was accepted by App Store Connect, but with the warning above, which would cause rejection if the app were to then be submitted for review.

### Reproduction Steps

Upload app to App Store Connect
Observe warning messages generated when upload is processed

(SDK team will provide support to retest this when API changes are made)

### Possible Solution

Use only public APIs on Apple frameworks

### aws-crt-swift version used

0.13.0

### Compiler and Version used

Xcode 15.0 / Swift 5.9

### Operating System and version

iOS (any)

---

**Comment by jbelkins** [comment]
The amplify-swift SDK has a related issue: https://github.com/aws-amplify/amplify-swift/issues/3324

**Comment by atierian** [comment]
Adding some findings from our (amplify-swift) investigation.

It appears that the addition of [common_cryptor_spi.h](https://github.com/awslabs/aws-c-cal/blob/main/source/darwin/common_cryptor_spi.h) is the culprit here - added in:
- https://github.com/awslabs/aws-c-cal/pull/136

I've only been able to reproduce the reported App Store rejection by:
1. distributing a DEBUG build.
-- or --
2. wrapping Amplify Swift (+ upstream: aws-sdk-swift, aws-crt-swift) in a framework, then linking that framework against my app target.

For 2., here's a simplified example:
**AwsCommonRuntimeKit --> MyApp**
```bash
> xcodebuild archive ...
> nm -m <archive-path>.xcarchive/Products/Applications/my_app.app/my_app | grep CCCryptorGCM # no results
```

**AwsCommonRuntimeKit --> wrapper.framework --> MyApp**
```bash
> xcodebuild archive ...
> nm -m <archive-path>.xcarchive/Products/Applications/my_app.app/Frameworks/wrapper.framework/wrapper | grep CCCryptorGCM
                 U _CCCryptorGCMAddAAD
                 U _CCCryptorGCMFinalize
                 U _CCCryptorGCMSetIV
```

Based on a dwarfdump of the framework, these symbols are coming from `aws-c-cal/source/darwin/common_cryptor_spi.h`.

```bash
> llvm-dwarfdump <archive-path>.xcarchive/dSYMs/wrapper.framework.dSYM | grep CCCryptorGCM -A 1
                DW_AT_name	("CCCryptorGCMSetIV")
                DW_AT_decl_file	("/.../checkouts/aws-crt-swift/aws-common-runtime/aws-c-cal/source/darwin/common_cryptor_spi.h")
--
                DW_AT_name	("CCCryptorGCMAddAAD")
                DW_AT_decl_file	("/.../checkouts/aws-crt-swift/aws-common-runtime/aws-c-cal/source/darwin/common_cryptor_spi.h")
--
                DW_AT_name	("CCCryptorGCMFinalize")
                DW_AT_decl_file	("/.../checkouts/aws-crt-swift/aws-common-runtime/aws-c-cal/source/darwin/common_cryptor_spi.h")

```

From briefly poking around aws-crt-swift, I'm only seeing two files that actually import `AwsCCal`; neither show an obvious code path to these SPI methods.
- [crt/ByteBuffer.swift](https://github.com/awslabs/aws-crt-swift/blob/main/Source/AwsCommonRuntimeKit/crt/ByteBuffer.swift#L6)
- [crt/Utilities.swift](https://github.com/awslabs/aws-crt-swift/blob/main/Source/AwsCommonRuntimeKit/crt/Utilities.swift#L6)

**Comment by graebm** [comment]
Fixed in 0.17.0
